src/coldcard.js

```diff
@@ -0,0 +1,431 @@
+/**
+ * Base class for file-based interactions with Coldcard
+ *
+ * @extends {module:coldcard.ColdcardInteraction}
+ */
+export class ColdcardFileReader extends ColdcardInteraction {
+
+  constructor({network}) {
+    super();
+    if ([MAINNET, TESTNET].find(net => net === network)) {
+      this.network = network;
+    } else {
+      throw new Error("Unknown network.");
+    }
+  }
+
+  messages() {
+    const messages = super.messages();
+    messages.push({
+      state: PENDING,
+      level: INFO,
+      code: "coldcard.upload",
+      text: "Upload the file from your Coldcard.",
+    });
+    return messages;
+  }
+
+  parse(file) {
+    //In the case of keys (json), the file will look like:
+    //
+    //{
+    //   "p2sh_deriv": "m/45'",
+    //   "p2sh": "tpubDA4nUAdTmY...MmtZaVFEU5MtMfj7H",
+    //   "p2wsh_p2sh_deriv": "m/48'/1'/0'/1'",
+    //   "p2wsh_p2sh": "Upub5THcs...Qh27gWiL2wDoVwaW",
+    //   "p2wsh_deriv": "m/48'/1'/0'/2'",
+    //   "p2wsh": "Vpub5n7tBWyvv...2hTzyeSKtZ5PQ1MRN",
+    //   "xfp": "12abcdef"
+    // }
+    //
+    // For now, we will derive unhardened from `p2sh_deriv`
+    // FIXME: assume we will gain the ability to ask Coldcard for an arbitrary path (or at least deeper+hardened)
+
+    if (typeof file === "object") {
+      this.xpubJSONFile = file;
+    }else if (typeof file === "string") {
+      try {
+        this.xpubJSONFile = JSON.parse(file);
+      } catch (error) {
+        throw new Error("Unable to parse JSON.");
+      }
+    } else {
+      throw new Error("Not valid JSON.");
+    }
+    if (Object.keys(this.xpubJSONFile).length === 0) {
+      throw new Error("Empty JSON file.");
+    }
+    return this.handleKeyExtraction(this.xpubJSONFile);
+    }
+
+  handleKeyExtraction(incomingJSON) {
+    let xpub = incomingJSON.p2sh;
+    const bip32Path = incomingJSON.p2sh_deriv;
+    const rootFingerprint = incomingJSON.xfp ? incomingJSON.xfp.toLowerCase() : null;
+
+    if (!xpub) {
+      throw new Error("No extended public key in JSON file.");
+    }
+    const xpubClass = ExtendedPublicKey.fromBase58(xpub);
+    let xfpFromWithinXpub;
+
+    if (!bip32Path) {
+      throw new Error("No BIP32 path in JSON file.");
+    }
+    if (!rootFingerprint && xpubClass.depth !== 1) {
+      throw new Error("No xfp in JSON file.");
+    }
+
+    /* istanbul ignore else */
+    // We can only find the fingerprint in the xpub if the depth is one
+    // because the xpub includes its parent's fingerprint.
+    if (xpubClass.depth === 1) {
+      xfpFromWithinXpub = fingerprintToFixedLengthHex(xpubClass.parentFingerprint).toLowerCase();
+    }
+
+    // Sanity check if you send in a depth one xpub, we should get the same fingerprint
+    if ((xfpFromWithinXpub && rootFingerprint) &&
+      xfpFromWithinXpub !== rootFingerprint) {
+      throw new Error("Computed fingerprint does not match the one in the file.");
+    }
+
+    return {
+      xpub,
+      rootFingerprint: rootFingerprint || xfpFromWithinXpub
```

Comment:
Yes - this method takes what was sent in via the JSON file and returns an object with both the `xpub` (renamed from JSON because it's listed as a different property) and adds on `rootFingerprint` which can either be included in the JSON (default) or computed from the xpub if it's depth==1 (bonus / sanity check)

This is specific to the format of the coldcard file so I'm unsure why it would go in `unchained-bitcoin` rather than here. The computation of xfp from depth1 xpub is using functionality from `unchained-bitcoin`.

Reply:
Yeah, the computation is what I meant. I guess what I'm asking then is why do we need a special method to rename a property and return a computed value that is computed elsewhere. If we need an object with these property names, a whole method seems superfluous imo. If the main point of this method is to sanity check all the values that the coldcard config needs, then couldn't it focus on just doing that and be named something like `validateConfigs`? if we still want a method to format that object, then separating out the validation from the object formatting feels appropriate.

Reply:
I think this may have been a holdover from when I was expecting a variety of uses from the `FileReader` instead of simply a single JSON file. I have edited for clarity.